next_sparseconvnet/networks/architectures.py

```diff
@@ -147,3 +147,118 @@ def forward(self, x):
         x = self.output(x)
         x = self.linear(x)
         return x
+
+
+class ResNet(torch.nn.Module):
+    '''
+    This class implements a net structure built with ResNet blocks. It takes a tuple of (coordinates, features)
+    and passes it through the ResNet.
+
+    ...
+
+    Attributes
+    ----------
+    spatial_size : tuple
+        The spatial size of the input layer. Size of the tuple is also the dimension.
+    inplanes : int
+        Number of planes we want after the initial SubmanifoldConvolution, that is, to begin downsampling.
+    kernel : list
+        Size of the kernels applied in each layer or block. Its length also indicates the level depth of the net.
+    stride : int
+        Applied stride in every layer or block.
+    conv_kernel : int
+        Kernel for the first convolutional layer.
+    basic_num : int
+        Number of times a basic residual block is passed in each level.
+    nclasses : int, optional
+        Number of output classes for predictions. The default is 2.
+    dim : int, optional
+        Number of dimensions of the input. The default is 3.
+    start_planes : int, optional
+        Number of planes that enter the ResNet. The default is 1.
+
+    Methods
+    -------
+    forward(x)
+        Passes the input through the ResNet
+    '''
+    def __init__(self, spatial_size, init_conv_nplanes, init_conv_kernel, kernel_sizes, stride, basic_num, nclasses = 2, dim = 3, start_planes = 1):
+        '''
+        Parameters
+        ----------
+        spatial_size : tuple
+            The spatial size of the input layer. Size of the tuple is also the dimension.
+        init_conv_nplaness : int
+            Number of planes we want after the initial SubmanifoldConvolution, that is, to begin downsampling.
+        init_conv_kernel : int
+            Kernel for the first convolutional layer.
+        kernel_sizes : list
+            Size of the kernels applied in each layer or block. Its length also indicates the level depth of the net.
+            Last element corresponds just to the kernel of the basic block in the bottom of the net.
+        stride : int
+            Applied stride in every layer or block.
+        basic_num : int
+            Number of times a basic residual block is passed in each level.
+        nclasses : int, optional
+            Number of output classes for predictions. The default is 2.
+        dim : int, optional
+            Number of dimensions of the input. The default is 3.
+        start_planes : int, optional
+            Number of planes that enter the net. The default is 1.
+        '''
+        torch.nn.Module.__init__(self)
+
+        self.basic_num   = basic_num
+        self.level_depth = len(kernel_sizes)
+
+        self.inp     = scn.InputLayer(dim, spatial_size)
+        self.convBN  = ConvBNBlock(start_planes, init_conv_nplanes, init_conv_kernel)
+        inplanes     = init_conv_nplanes
+        self.max     = scn.MaxPooling(dim,2,2)
```

Comment:
The MaxPool layer has pool size and stride, we probably want to do _global_ maxpool, ie, you calculate the final spatial dimension and use it as pool size